data-generation/generate-instructions.py

```diff
@@ -0,0 +1,204 @@
+            (self.side1_letter, self.side1_color,
+             self.side2_letter, self.side2_color,
+             self.position)
+
+
+class Instruction:
+    def __init__(self, phrase, point):
+        self.phrase = phrase
+        self.point = point
+
+    def __str__(self):
+        return self.phrase
+
+
+class Configuration:
+    def __init__(self, blocks, final_blocks=[]):
+        self.current_blocks = blocks
+        self.final_blocks = final_blocks
+
+    def is_complete(self):
+        return self.current_blocks == []
+
+    def get_instruction(self, moved_block):
+        point = moved_block.position
+        phrase = 'move the %s %s block here: %s' \
+                 % (moved_block.side1_color,
+                    moved_block.side1_letter,
+                    point)
+        return Instruction(phrase, point)
+
+    def get_action(self, goal):
+        if self.is_complete():
+            raise NoActionException('Board is already complete')
+
+        block_to_move = rand_element(self.current_blocks)
+        moved_block = goal.final_blocks[goal.final_blocks.index(block_to_move)]
+        new_current_blocks = self.current_blocks[:]
+        new_current_blocks.remove(moved_block)
+        new_configuration = Configuration(
+            new_current_blocks,
+            self.final_blocks + [moved_block]
+        )
+        instruction = self.get_instruction(moved_block)
+        return Action(self, new_configuration, instruction)
+
+    def scatter(self):
+        return Configuration(list(map(randomize_block,
+                                      self.current_blocks + self.final_blocks)))
+
+    def __str__(self):
+        # :(
+        return str(list(map(str, self.current_blocks + self.final_blocks)))
+
+    def mark_complete(self):
+        return FinalConfiguration(self.current_blocks + self.final_blocks)
+
+
+class FinalConfiguration(Configuration):
+    def __init__(self, blocks):
+        self.current_blocks = []
+        self.final_blocks = blocks
+
+
+class Action:
+    def __init__(self, start_conf, end_conf, phrase):
+        self.start_conf = start_conf
+        self.end_conf = end_conf
+        self.phrase = phrase
+
+    def __str__(self):
+        return str(self.phrase)
+
+
+"""
+Randomness
+"""
+
+def rand_element(ls):
+    return ls[rand_index(ls)]
+
+
+def rand_index(ls):
+    return randint(0, len(ls) - 1)
+
+
+def random_position():
+    return (randint(0, 1000), randint(0, 1000))
+
+
+def random_block(letters, colors):
+    side1_letter = rand_element(letters)
+    side2_letter = rand_element(letters)
+    side1_color  = rand_element(colors)
+    side2_color  = rand_element(colors)
+    position = random_position()
+    return Block(
+        side1_letter,
+        side1_color,
+        side2_letter,
+        side2_color,
+        position
+    )
+
+
+def randomize_block(block):
+    def maybe_flip(b):
+        # if random() < 0.0:
+        #     return b.flip()
+        return b
```

Comment:
What's going on here?

Reply:
I haven't implemented block flipping for the "solving" part so I just commented it out here. I left it there just because that is exactly what the function will look like when that is implemented. Sorry for not making that a little more clear

Reply:
👍